### PR TITLE
[SPARK-7319][SQL] Improve the output from DataFrame.show()

### PR DESCRIPTION
**Pull request by kaka1992**

### CODE REVIEW

**Comment by marmbrus** [comment]
Please include a before and after example of the output and a description of what you changed (in the PR description).

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -592,10 +592,10 @@ class DataFrameSuite extends QueryTest {
     checkAnswer(df.select(df("key")), testData.select('key).collect().toSeq)
   }
 
-  ignore("show") {
+  test("show") {
     // This test case is intended ignored, but to make sure it compiles correctly
     testData.select($"*").show()
-    testData.select($"*").show(1000)
+    //testData.select($"*").show(1000)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Undo this change.

**Comment by kaka1992** [comment]
testData.select($"*").show(5)

Before:
key value
1   1    
2   2    
3   3    
4   4    
5   5 

After:
+----+-----+
|key|value|
|    1|      1|
|    2|      2|
|    3|      3|
|    4|      4|
|    5|      5|
+----+------+

**Comment by rxin** [comment]
@kaka1992 Thanks a lot for being so helpful. I have two requests:

1. Can you add a line after the header, exactly like what I wrote in the JIRA ticket?
2. Can you read this coding style guide and clean up the code? https://github.com/databricks/scala-style-guide  In general, it would be great to have the style match exactly rest of the code. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -175,6 +175,7 @@ class DataFrame private[sql](
    * @param numRows Number of rows to show
    */
   private[sql] def showString(numRows: Int): String = {
+    val sb = new StringBuilder
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: it would be great to create unit test for showString in DataFrameSuite.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 3]
```diff
@@ -194,12 +195,34 @@ class DataFrame private[sql](
...
-    rows.map { row =>
-      row.zipWithIndex.map { case (cell, i) =>
-        String.format(s"%-${colWidths(i)}s", cell)
-      }.mkString(" ")
-    }.mkString("\n")
+    // Create SeparateLine
+    val sep:String = {
+      "+" + colWidths.map {
+        size =>
+          val columnSep = new Array[Char](size)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: this whole block is really just 
```scala
"-" * size
```

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 4]
```diff
@@ -194,12 +195,34 @@ class DataFrame private[sql](
...
+            columnSep(i) = '-'
+          String.valueOf(columnSep)
+      }.mkString("+") + "+\n"
+    }
+
+    sb.append(sep)
+
+    // append column names
+    sb.append("|").append(rows.head.zipWithIndex.map { case (cell, i) =>
+        String.format(s"%${colWidths(i)}s", cell)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: maybe use the padding function on string directly, rather than doing the format.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 5]
```diff
@@ -194,12 +195,34 @@ class DataFrame private[sql](
...
+      }.mkString("|")).append("|\n")
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: indentation here is off, and it is very confusing to have it wrapping this way

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 6]
```diff
@@ -598,6 +598,11 @@ class DataFrameSuite extends QueryTest {
     testData.select($"*").show(1000)
   }
 
+  test("SPARK-7319 showString") {
+    assert(testData.select($"*").showString(1).split("\n") === Seq("+---+-----+",
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: you can use multi line string in Scala, especially with stripMargin.

http://alvinalexander.com/scala/scala-multiline-strings-heredoc-syntax

**Comment by rxin** [comment]
This is looking pretty good now. I think you would need to fix some Python tests as well, as they were expecting the output from show.

**Comment by kaka1992** [comment]
@rxin Please test this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 7]
```diff
@@ -598,6 +598,23 @@ class DataFrameSuite extends QueryTest {
     testData.select($"*").show(1000)
   }
 
+  test("SPARK-7319 showString") {
+    assert(testData.select($"*").showString(1) === """+---+-----+
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: it's pretty awkward to have this at the end. I'd just create a variable for the expected string so we don't have so much whitespace.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by kaka1992** [comment]
@rxin error: oro#oro;2.0.8!oro.jar origin location must be absolute. What's wrong with building?

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by kaka1992** [comment]
@rxin Fixed showDF test in R.

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by shivaram** [comment]
SparkR changes look good

**Comment by rxin** [comment]
Thanks a lot for doing this. I'm merging this in master.
